src/index.ts

```diff
@@ -0,0 +1,82 @@
+import {
+	AbstractSqlField,
+	AbstractSqlModel,
+	AbstractSqlTable,
+} from '@balena/abstract-sql-compiler';
+import { sqlNameToODataName } from '@balena/odata-to-abstract-sql';
+import { replaceResultTransformer, TemplateTag } from 'common-tags';
+
+const trimNL = new TemplateTag(
+	replaceResultTransformer(/^[\r\n]*|[\r\n]*$/g, ''),
+);
+
+const modelNameToTypescriptName = (s: string) =>
+	s
+		.split(/[ -]/)
+		.map((p) => p[0].toLocaleUpperCase() + p.slice(1))
+		.join('');
+
+const sqlTypeToTypescriptType = (m: AbstractSqlModel, f: AbstractSqlField) => {
+	switch (f.dataType) {
+		case 'Boolean':
+			return 'boolean';
+		case 'Short Text':
+		case 'Text':
+		case 'Hashed':
+			return 'string';
+		case 'Date Time':
+			return 'Date';
+		case 'Serial':
+		case 'Integer':
+		case 'Big Integer':
+		case 'Real':
+			return 'number';
+		// TODO: ForeignKey/ConceptType should really be that of the referenced id
+		case 'ForeignKey':
+		case 'ConceptType':
+			const referencedInterface = modelNameToTypescriptName(
+				m.tables[f.references!.resourceName].name,
+			);
+			const nullable = f.required ? '' : '?';
+			return `{ __id: number } | [${referencedInterface}${nullable}]`;
```

Comment:
I think that might be an issue in the SDK typings @thgreasi and how it does typing magic since `[T?]` vs `[] | [T]` should be equivalent and at least on a basic test using
```ts
const x: OptionalNavigationResource<Device> = [] as any;
if (Array.isArray(x)) {
	const y = x[0]
}
```
then `y` ends up with the same type with either `[T?]` or `[] | [T]`

Reply:
Well strictly speaking `[T?]` means that there is a `T` or `undefined` in the single item tuple, while the other describes that it's either a single item `T` tuple or an empty array.

Reply:
`T[0]` will return `undefined` in both cases @thgreasi and check the following:
```ts
let x: [string?]
type l = typeof x.length
// type l = 0 | 1
```
the typing seems to recognise it's either length 0 or 1, the same as for `[] | [T]`:
```ts
let x: [] | [string]
type l = typeof x.length
// type l = 0 | 1